### DISCUSSION TITLE
Magnitisation in first limiting case

### DISCUSSION THREAD

**Discussion by AprilSweettooth** [Part-II-Computational-Physics/.github#59] [Q&A]
I am not sure why I could not get the lower half matched with the analytical value... Is it because my timestamp is not right, I used avg_times = [10,100,1000,10000] for this one.
![image](https://user-images.githubusercontent.com/72824407/223075310-03f3cc4f-b31c-4bb0-a0f5-b95a025fe537.png)

**Comment by DLHvdH** [comment]
It seems like the negative half of the curve isn't working well. Are you sure the signs of the magnetisation are correct? The analytical expression considers both negative and positive magnetisations, giving an expression that ranges between -1 and +1.

**Reply by AprilSweettooth** [reply]
I think so? 

'betaMuHs_analytical = np.linspace(-3,3,100) # betaMuH values to plot
M_analytical = np.tanh(betaMuHs_analytical)

avg_times = [10,100,1000,10000]
lattice=init_lattice(12, type=0)
print(lattice)
M = []
for b in betaMuHs_analytical:
  M.append(np.average(evolve_and_compute_M(lattice,0,b,avg_times)))
plt.plot(betaMuHs_analytical, M, 'r', label='Monte Carlo')
plt.plot(betaMuHs_analytical, M_analytical, 'black', label='Analytical')
plt.xlabel ('Magnetic field strength betaMuH')
plt.ylabel ('Magnetisation M')
plt.legend(loc='upper left')
plt.show() '

I just used (1/len(lattice)**2) * np.sum(lattice) for my magnetisation. But it seems the negatiev half only concentrate on -0.25.

**Comment by PatrickC12** [comment]
Does your evolve_and_compute_M() average M over all times in avg_times ? Perhaps the system hasn't had enough time steps to thermalise yet at some of those values of avg_times.

edit: although I'm not sure why MC would only work well for large positive betaMuH then, unless your initial lattice is biased

**Reply by AprilSweettooth** [reply]
I think so, that's why I am not sure about which avg times to pick...

**Reply by PatrickC12** [reply]
I used:

avg_times = np.arange(0,100000,100)

I think the system thermalises in around 10^4 steps. So I made sure to only average M for t>10^4 in evolve_and_compute_M()

edit: takes a while to calculate and plot, but plot works very well

**Reply by AprilSweettooth** [reply]
Thanks. i think this indeed can smooth the curve, but i still got the bias somehow...
![image](https://user-images.githubusercontent.com/72824407/223207499-0196056e-5dba-45c8-bf7a-34beb2d51f24.png)

**Reply by PatrickC12** [reply]
Yeah, not sure about that. I would have to see  your evolve_and_compute code perhaps. It looks like you used a random lattice in your code so there shouldn't be any of this bias

**Reply by AprilSweettooth** [reply]
def evolve_and_compute_M(lattice,betaJ,betaMuH,avg_times):
    '''Evolves the lattice using the Metropolis-Hastings algorithm and returns the average magnetisation computed using different time steps.'''
    # TO DO
    magnetisation = []
    for t in range(avg_times[-1]+1):
        if t in avg_times:
            magnetisation.append(compute_magnetisation(lattice))
        lattice = attempt_spin_flip(lattice, len(lattice), betaJ, betaMuH)
    return magnetisation

This is what I have used

**Reply by PatrickC12** [reply]
I mean that looks fine to me. I guess you could put extra if statement(only average equilibria):

if t in avg_times:
          if t >10000:
                magnetisation.append(compute_magnetisation(lattice))

But this shouldn't effect mean too much if it only effects ~10% of np.arange(0,100000,100).

I can't really see what would be causing it, assuming you got the expected behaviour with your attempt_spin_flip() in the previous sections

**Reply by AprilSweettooth** [reply]
Thanks for reminding me that! I just realized I forgot to update my spin parameter in the attemp func.
![image](https://user-images.githubusercontent.com/72824407/223222129-aefcab6d-cc2d-405c-998a-254582b70979.png)

All good now

**Reply by PatrickC12** [reply]
Looking better, not sure about the behaviour at lower field strengths there, perhaps your system hasn't come to equilibrium before averaging has occurred.

I got this plot
![image](https://user-images.githubusercontent.com/123903514/223223910-12079bc9-7a86-44cf-80cc-caff8aa9fcfa.png)

**Reply by AprilSweettooth** [reply]
Yeah, I changed the avg_times because it takes too long to run

**Reply by PatrickC12** [reply]
Ah fair , I included a counter in my for loop mine just so I could leave it running in the background and check it was still working away

**Reply by AprilSweettooth** [reply]
good idea

**Reply by benpymer** [reply]
Wait so, are we meant to be calculating the average magnetisation of one run of the simulation over multiple times for each value of betamuh, or are we calculating the average magnetisation at certain sample points in time over multiple runs of the simulation for each value of betamuh? I had originally interpreted the task as the latter, but I think it would take an exceptionally long time to do and also this thread seems to be talking as if it is the former?

**Reply by DLHvdH** [reply]
You can either average the magnetisation of one simulation over multiple points in time. Or you can do multiple simulations **at the same betaMuH** and average their magnetisations. You will have to do either to get a stable average magnetisation. The former is actually faster. You can figure out why if you think about what the intitial lattice looks like.

**Reply by AprilSweettooth** [reply]
```
def attempt_spin_flip(lattice,width,betaJ,betaMuH):
    '''Applies the Metropolis-Hastings algorithm to try and flip a spin.'''
    # TO DO
    w = len(lattice)
    i,j = rng.integers(0,w,2)
    if lattice[i,j] * neighbouring_spins_sum(i,j,lattice,width) <= 0 or rng.random() < np.exp(-1 * compute_betaDeltaE(i,j,lattice,width,betaJ,betaMuH)):
        lattice[i,j] *= -1
    return lattice
def evolve_and_plot(lattice,betaJ,betaMuH,plot_times):
    '''Evolves the lattice using the Metropolis-Hastings algorithm and plots the lattice at different times.'''
    # TO DO
    fig, ax = plt.subplots(1,len(plot_times),figsize=(12,4))
    for t in range(plot_times[-1]+1):
        # TO DO
        if t in plot_times:
            plot_lattice(lattice,ax[plot_times.index(t)],"t = {}".format(t))
        lattice = attempt_spin_flip(lattice,len(lattice),betaJ,betaMuH)
    plt.show()

def evolve_and_compute_M(lattice,betaJ,betaMuH,avg_times):
    '''Evolves the lattice using the Metropolis-Hastings algorithm and returns the average magnetisation computed using different time steps.'''
    # TO DO
    magnetisation = []
    for t in range(avg_times[-1]+1):
        if t in avg_times:
            magnetisation.append(compute_magnetisation(lattice))
        lattice = attempt_spin_flip(lattice, len(lattice), betaJ, betaMuH)
    return magnetisation

betaMuHs_analytical = np.linspace(-3,3,100) # betaMuH values to plot
M_analytical = np.tanh(betaMuHs_analytical)

avg_times = np.arange(10000,100000,1000)
# Pick some suitable time points at which to sample your magnetisation. These values should be averaged to get a stable mean

# TO DO: Simulate the system at different magnetic field strengths
lattice=init_lattice(12, type=0)
print(lattice)
M = []
for b in betaMuHs_analytical:
  M.append(np.average(evolve_and_compute_M(lattice,0,b,avg_times)))
plt.plot(betaMuHs_analytical, M, 'r', label='Monte Carlo')
plt.plot(betaMuHs_analytical, M_analytical, 'black', label='Analytical')
plt.xlabel ('Magnetic field strength betaMuH')
plt.ylabel ('Magnetisation M')
plt.legend(loc='upper left')
plt.title('Comparing Monte Carlo to analytical solution of Magnetisation with zero coupling')
plt.show()
```

Turns out I still got the dips in the simulation even if I changed to longer avg_times... I am not sure why.

Also I think the same thing affects my limit case 2,

```
# # TO DO
# in T basis
# k_b = 2*J for ferro and -2*J for anti
Tc = 1/np.log(1+np.sqrt(2))
T_analytical = np.linspace(0.01*Tc,2*Tc,200)
M_analytical_ferro = np.zeros((200,))
for i in range(100):
  M_analytical_ferro[i] = (1-np.sinh(1/T_analytical[i])**(-4))**(1/8)
M_analytical_anti = -M_analytical_ferro # generate both ferromagnetic and antiferro at the same time

avg_times = np.arange(1000,10000,1000)
# Pick some suitable time points at which to sample your magnetisation. These values should be averaged to get a stable mean

# TO DO: Simulate the system at different magnetic field strengths
lattice_up=init_lattice(15, type=1)
lattice_anti=np.resize([1,-1], (15,15))
M_ferro = []
M_anti = []
for T in T_analytical:
  M_ferro.append(np.average(evolve_and_compute_M(lattice_up,1/T,0,avg_times)))
  M_anti.append(np.average(evolve_and_compute_M(lattice_anti,-1/T,0,avg_times)))
plt.plot(T_analytical, M_ferro, 'y', label='Monte Carlo ferro')
plt.plot(T_analytical, M_anti, 'b', label='Monte Carlo anti')
plt.plot(T_analytical, M_analytical_ferro, 'black', label='ferromagnetic Analytical')
plt.plot(T_analytical, M_analytical_anti, 'red', label='antiferromagnetic Analytical')
# Simuatle the system at different coupling strengths
plt.title('Magnetisation with only coupling')
plt.xlabel ('Temperature over critical temperature')
plt.ylabel ('Magnetisation M')
plt.legend(loc='upper right')
plt.show()
```

Because I got something like this:
![image](https://user-images.githubusercontent.com/72824407/223749548-00dbbe4e-f17d-4198-a755-534aa6bdf755.png)

I mean, the computation time is indeed faster with right guess. But the curve is just way off...
